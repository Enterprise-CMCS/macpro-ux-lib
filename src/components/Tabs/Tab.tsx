import React from "react";

type IntrinsicElements = JSX.IntrinsicElements["a"];

interface TabProps extends IntrinsicElements {
  disabled?: boolean;
  tabLabel: string;
}

/**
 * **Tab Component**
 *
 * Used as a helper for the Tabs component. Not intended to be used as a standalone resource.
 *
 */
export const Tab: React.FC<TabProps> = ({
  className,
  disabled,
  tabLabel,
  onClick,
  ...rest
}) => {
  return disabled ? (
    <span className={`${className} disabled`} aria-disabled="true" {...rest}>
      {tabLabel}
    </span>
  ) : (
    <a className={className} onClick={onClick} {...rest}>
      {tabLabel}
    </a>
  );
};
