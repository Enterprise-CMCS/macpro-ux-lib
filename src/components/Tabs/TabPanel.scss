@use "../../assets/theme/macbis-colors" as c;

// .tabPanel {
.tab {
  border: 1px solid c.$gray-lighter;
  color: c.$base;
  font-size: 16px;
  font-weight: 700;
  margin: 0 -1px -1px 0;
  padding: 16px 24px;
  text-decoration: none;
  &:hover {
    color: c.$primary;
  }
  &.current {
    border-bottom: 1px solid white;
    color: c.$primary;
    position: relative;
    &::after {
      background-color: c.$primary;
      content: "";
      height: 4px;
      left: -1px;
      position: absolute;
      right: -1px;
      top: -1px;
    }
  }
}
// }
