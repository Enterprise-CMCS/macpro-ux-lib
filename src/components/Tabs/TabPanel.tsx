import React, { PropsWithChildren } from "react";

type IntrinsicElements = JSX.IntrinsicElements["h4"];

export interface TabPanelProps extends IntrinsicElements {
  disabled?: boolean;
  id: string;
  tabLabel: string;
}

/**
 * **TabPanel Component**
 *
 * The `TabPanel` is intended to be used as a child of `Tabs`. Each `TabPanel` represents a tab and the corresponding content
 * to be rendered when the tab is active.
 *
 * ```html
 * <Tabs>
 *   <TabPanel tab="Summary">The Bill of...</TabPanel>
 * </Tabs>
 * ```
 *
 * @param {React.Node}    children  Contents to be displayed when Tab is clicked.
 * @param {boolean}       disabled  Setting this value will render Tab with disabled styles and make tab unclickable.
 * @param {string}        id        Unique identifier required for each TabPanel item. The id is used to associate the tab and it's content, as well as populate aria tags.
 * @param {string}        tabLabel  Text label for each tab.
 *
 */
export const TabPanel: React.FC<PropsWithChildren<Props>> = ({
  children,
  id,
  tabLabel,
  ...rest
}) => {
  return (
    <div {...rest} aria-labelledby={id} role="tabpanel">
      {children}
    </div>
  );
};
