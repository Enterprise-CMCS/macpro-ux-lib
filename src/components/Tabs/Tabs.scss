@use "../../assets/theme/macbis-colors" as c;

.tabs {
  .tabs-row {
    border-bottom: 1px solid c.$gray-lighter;
    display: flex;
    flex-direction: row;
  }
  .tabs-content {
    border: 1px solid c.$gray-lighter;
    border-top: none;
    padding: 24px;
  }
}
