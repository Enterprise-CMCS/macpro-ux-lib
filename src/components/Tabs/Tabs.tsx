import React, { Children, PropsWithChildren, useState } from "react";
import { Tab } from "./Tab";

type IntrinsicElements = JSX.IntrinsicElements["div"];

export interface TabsProps extends IntrinsicElements {
  initialTab?: number;
}

/**
 * **Tabs Component**
 *
 * `Tabs` acts as a parent for one or many `TabPanel` children. These components are reliant on
 * each other and will not display correctly if used independently.
 *
 * ```html
 * <Tabs>
 *   <TabPanel></TabPanel>
 *   <TabPanel></TabPanel>
 *   <TabPanel></TabPanel>
 * </Tabs>
 * ```
 *
 * @param {React.Node}    children    `TabPanel` children to be rendered in the Tabs component.
 * @param {number}        initialTab  Index of the tab to be open when the component renders. Defaults to first non-disabled tab.
 */
export const Tabs: React.FC<PropsWithChildren<TabsProps>> = ({
  children,
  initialTab,
  ...rest
}) => {
  const arrayChildren = Children.toArray(children);

  const getInitialTab = () => {
    for (const child of arrayChildren) {
      if (React.isValidElement(child) && child.props.disabled !== true) {
        return arrayChildren.indexOf(child);
      }
    }
    return null;
  };

  // State of all tab content
  const [currentTab, setCurrentTab] = useState<number | null>(
    initialTab || getInitialTab()
  );

  return (
    <div className="tabs" {...rest}>
      <div className="tabs-row" role="tablist">
        {Children.map(arrayChildren, (child, idx) => {
          if (React.isValidElement(child)) {
            const current = currentTab === idx;
            const classNames = `tab${current ? " current" : ""}`;
            return (
              <Tab
                aria-controls={child.props.tab}
                aria-selected={current}
                aria-hidden={!current}
                {...child.props}
                className={classNames}
                id={child.props.id}
                onClick={() => setCurrentTab(idx)}
                role="tab"
              />
            );
          }
        })}
      </div>
      <div className="tabs-content">
        {Children.map(arrayChildren, (child, idx) => {
          if (React.isValidElement(child)) {
            const current = currentTab === idx;
            return React.cloneElement(child as React.ReactElement, {
              hidden: !current,
              "aria-hidden": !current,
            });
          }
        })}
      </div>
    </div>
  );
};
