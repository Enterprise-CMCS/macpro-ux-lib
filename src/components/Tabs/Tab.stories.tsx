import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Tab } from "./Tab";
import { TabPanel } from "./TabPanel";

export default {
  title: "COMPONENTS/Tabs/Tab",
  component: Tab,
  args: {},
  argTypes: {},
} as ComponentMeta<typeof Tab>;

const Template: ComponentStory<typeof Tab> = ({ ...rest }) => <Tab {...rest} />;

export const Default = Template.bind({});
Default.args = {};
