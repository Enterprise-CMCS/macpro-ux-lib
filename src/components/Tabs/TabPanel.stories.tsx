import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { TabPanel } from "./TabPanel";
import { Tabs } from "./Tabs";

export default {
  title: "COMPONENTS/Tabs/Tab",
  component: TabPanel,
  args: {},
  argTypes: {},
} as ComponentMeta<typeof TabPanel>;

const Template: ComponentStory<typeof TabPanel> = ({ ...rest }) => (
  <TabPanel {...rest} />
);

export const Default = Template.bind({});
Default.args = {};
