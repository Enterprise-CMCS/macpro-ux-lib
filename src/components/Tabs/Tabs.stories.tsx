import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { TabPanel } from "./TabPanel";
import { Tabs } from "./Tabs";

export default {
  title: "COMPONENTS/Tabs/Tabs",
  component: Tabs,
  args: {
    children: "",
    initialTab: undefined,
  },
  argTypes: {
    children: {
      description: "`TabPanel` children to be rendered in the Tabs component.",
      type: "function",
    },
    initialTab: {
      description:
        "Index of the tab to be open when the component renders. Defaults to first non-disabled tab.",
    },
  },
} as ComponentMeta<typeof Tabs>;

const Template: ComponentStory<typeof Tabs> = ({ ...rest }) => (
  <Tabs {...rest}>
    <TabPanel tabLabel="Summary" id="tab-panel--summary">
      The Bill of Rights is the first ten amendments to the United States
      Constitution.
    </TabPanel>
    <TabPanel tabLabel="Preamble" id="tab-panel--preamble">
      We the People of the United States, in Order to form a more perfect Union,
      establish Justice, insure domestic Tranquility, provide for the common
      defence, promote the general Welfare, and secure the Blessings of Liberty
      to ourselves and our Posterity, do ordain and establish this Constitution
      for the United States of America.
    </TabPanel>
    <TabPanel tabLabel="Amendments" id="tab-panel--amendments">
      <h2>Bill of Rights</h2>
      <ol>
        <li>Freedoms, Petitions, Assembly</li>
        <li>Right to bear arms</li>
        <li>Quartering of soldiers</li>
        <li>Search and arrest</li>
        <li>Rights in criminal cases</li>
        <li>Right to a fair trial</li>
        <li>Rights in civil cases</li>
        <li>Bail, fines, punishment</li>
        <li>Rights retained by the People</li>
        <li>States' rights</li>
      </ol>
      <h2>Later Amendments</h2>
      <ol start={11}>
        <li>Lawsuits against states</li>
        <li>Presidential elections</li>
        <li>Abolition of slavery</li>
        <li>Civil rights</li>
        <li>Black suffrage</li>
        <li>Income taxes</li>
        <li>Senatorial elections</li>
        <li>Prohibition of liquor</li>
        <li>Women's suffrage</li>
        <li>Terms of office</li>
        <li>Repeal of Prohibition</li>
        <li>Term Limits for the Presidency</li>
        <li>Washington, D.C., suffrage</li>
        <li>Abolition of poll taxes</li>
        <li>Presidential succession</li>
        <li>18-year-old suffrage</li>
        <li>Congressional pay raises</li>
      </ol>
    </TabPanel>
    <TabPanel tabLabel="Disabled" disabled id="tab-panel--disabled" />
  </Tabs>
);

export const Default = Template.bind({});
Default.args = {};
