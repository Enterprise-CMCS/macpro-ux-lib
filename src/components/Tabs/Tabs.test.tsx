import React from "react";
import { fireEvent, screen, render } from "../../test-setup";
import { TabPanel } from "./TabPanel";
import { Tabs } from "./Tabs";

const tabPanels = [
  <TabPanel tab="Summary" id="tab-panel--summary">
    The Bill of Rights is the first ten amendments to the United States
    Constitution.
  </TabPanel>,
  <TabPanel tab="Preamble" id="tab-panel--preamble">
    We the People of the United States, in Order to form a more perfect Union,
    establish Justice, insure domestic Tranquility, provide for the common
    defence, promote the general Welfare, and secure the Blessings of Liberty to
    ourselves and our Posterity, do ordain and establish this Constitution for
    the United States of America.
  </TabPanel>,
  <TabPanel tab="Amendments" id="tab-panel--amendments">
    <h2>Bill of Rights</h2>
    <ol>
      <li>Freedoms, Petitions, Assembly</li>
      <li>Right to bear arms</li>
      <li>Quartering of soldiers</li>
      <li>Search and arrest</li>
      <li>Rights in criminal cases</li>
      <li>Right to a fair trial</li>
      <li>Rights in civil cases</li>
      <li>Bail, fines, punishment</li>
      <li>Rights retained by the People</li>
      <li>States' rights</li>
    </ol>
    <h2>Later Amendments</h2>
    <ol start={11}>
      <li>Lawsuits against states</li>
      <li>Presidential elections</li>
      <li>Abolition of slavery</li>
      <li>Civil rights</li>
      <li>Black suffrage</li>
      <li>Income taxes</li>
      <li>Senatorial elections</li>
      <li>Prohibition of liquor</li>
      <li>Women's suffrage</li>
      <li>Terms of office</li>
      <li>Repeal of Prohibition</li>
      <li>Term Limits for the Presidency</li>
      <li>Washington, D.C., suffrage</li>
      <li>Abolition of poll taxes</li>
      <li>Presidential succession</li>
      <li>18-year-old suffrage</li>
      <li>Congressional pay raises</li>
    </ol>
  </TabPanel>,
  <TabPanel tab="Disabled" disabled id="tab-panel--disabled" />,
];

describe("Tests for the Tabs component", () => {
  it("should render", () => {
    const { container } = render(<Tabs children={tabPanels} />);

    // verify content is displayed as expected
    const tabs = container.getElementsByClassName("tab");
    expect(tabs.length).toBe(4);
    const currentTab = container.getElementsByClassName("current");
    expect(currentTab.length).toBe(1);
    expect(currentTab[0].innerHTML).toBe("Summary");
    const disabledTab = container.getElementsByClassName("disabled");
    expect(disabledTab.length).toBe(1);
    expect(disabledTab[0].innerHTML).toBe("Disabled");
    const tabsContent = container.getElementsByClassName("tabs-content");
    expect(tabsContent.length).toBe(1);
    const tabPanelContent = Array.from(tabsContent[0].children);
    expect(tabPanelContent.length).toBe(4);
    const currentContent = tabPanelContent.filter(
      (elem) => elem.getAttribute("aria-hidden") !== "true"
    );
    expect(currentContent.length).toBe(1);
    expect(currentContent[0].innerHTML).toBe(
      "The Bill of Rights is the first ten amendments to the United States Constitution."
    );
  });

  it("clicking a tab should set it to current", () => {
    const { container } = render(<Tabs children={tabPanels} />);
    expect(container.getElementsByClassName("current")[0].innerHTML).toBe(
      "Summary"
    );

    const preamble = screen.getByText("Preamble");
    fireEvent.click(preamble);
    expect(container.getElementsByClassName("current")[0].innerHTML).toBe(
      "Preamble"
    );
    const tabPanelContent =
      container.getElementsByClassName("tabs-content")[0].children;
    const currentContent = Array.from(tabPanelContent).filter(
      (elem) => elem.getAttribute("aria-hidden") !== "true"
    );
    expect(currentContent.length).toBe(1);
    expect(currentContent[0].innerHTML).toBe(
      "We the People of the United States, in Order to form a more perfect Union, establish Justice, insure domestic Tranquility, provide for the common defence, promote the general Welfare, and secure the Blessings of Liberty to ourselves and our Posterity, do ordain and establish this Constitution for the United States of America."
    );
  });

  it("clicking a disabled tab should do nothing", () => {
    const { container } = render(<Tabs children={tabPanels} />);
    expect(container.getElementsByClassName("current")[0].innerHTML).toBe(
      "Summary"
    );
    const disabled = screen.getByText("Disabled");
    fireEvent.click(disabled);
    expect(container.getElementsByClassName("current")[0].innerHTML).toBe(
      "Summary"
    );
  });

  it("providing an initial tab index sets the current tab", () => {
    const { container } = render(<Tabs children={tabPanels} initialTab={1} />);
    expect(container.getElementsByClassName("current")[0].innerHTML).toBe(
      "Preamble"
    );
    const tabPanelContent =
      container.getElementsByClassName("tabs-content")[0].children;
    const currentContent = Array.from(tabPanelContent).filter(
      (elem) => elem.getAttribute("aria-hidden") !== "true"
    );
    expect(currentContent.length).toBe(1);
    expect(currentContent[0].innerHTML).toBe(
      "We the People of the United States, in Order to form a more perfect Union, establish Justice, insure domestic Tranquility, provide for the common defence, promote the general Welfare, and secure the Blessings of Liberty to ourselves and our Posterity, do ordain and establish this Constitution for the United States of America."
    );
  });

  it("when all tabs are disabled none should be set to current", () => {
    const { container } = render(
      <Tabs>
        <TabPanel tab="Disabled 1" disabled id="tab-panel--disabled-1" />
        <TabPanel tab="Disabled 2" disabled id="tab-panel--disabled-2" />
        <TabPanel tab="Disabled 3" disabled id="tab-panel--disabled-3" />
      </Tabs>
    );
    expect(container.getElementsByClassName("current").length).toBe(0);
  });

  it("component snapshot", () => {
    const { container } = render(<Tabs children={tabPanels} />);
    expect(container).toMatchSnapshot();
  });
});
