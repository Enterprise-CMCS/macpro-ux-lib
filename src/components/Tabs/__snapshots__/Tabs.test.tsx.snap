// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the Tabs component component snapshot 1`] = `
<div>
  <div
    class="tabs"
  >
    <div
      class="tabs-row"
      role="tablist"
    >
      <a
        aria-hidden="false"
        aria-selected="true"
        class="tab current"
        id="tab-panel--summary"
        role="tab"
      >
        Summary
      </a>
      <a
        aria-hidden="true"
        aria-selected="false"
        class="tab"
        id="tab-panel--preamble"
        role="tab"
      >
        Preamble
      </a>
      <a
        aria-hidden="true"
        aria-selected="false"
        class="tab"
        id="tab-panel--amendments"
        role="tab"
      >
        Amendments
      </a>
      <span
        aria-disabled="true"
        aria-hidden="true"
        aria-selected="false"
        class="tab disabled"
        id="tab-panel--disabled"
        role="tab"
      >
        Disabled
      </span>
    </div>
    <div
      class="tabs-content"
    >
      <div
        aria-hidden="false"
        aria-labelledby="tab-panel--summary"
        role="tabpanel"
      >
        The Bill of Rights is the first ten amendments to the United States Constitution.
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="tab-panel--preamble"
        hidden=""
        role="tabpanel"
      >
        We the People of the United States, in Order to form a more perfect Union, establish Justice, insure domestic Tranquility, provide for the common defence, promote the general Welfare, and secure the Blessings of Liberty to ourselves and our Posterity, do ordain and establish this Constitution for the United States of America.
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="tab-panel--amendments"
        hidden=""
        role="tabpanel"
      >
        <h2>
          Bill of Rights
        </h2>
        <ol>
          <li>
            Freedoms, Petitions, Assembly
          </li>
          <li>
            Right to bear arms
          </li>
          <li>
            Quartering of soldiers
          </li>
          <li>
            Search and arrest
          </li>
          <li>
            Rights in criminal cases
          </li>
          <li>
            Right to a fair trial
          </li>
          <li>
            Rights in civil cases
          </li>
          <li>
            Bail, fines, punishment
          </li>
          <li>
            Rights retained by the People
          </li>
          <li>
            States' rights
          </li>
        </ol>
        <h2>
          Later Amendments
        </h2>
        <ol
          start="11"
        >
          <li>
            Lawsuits against states
          </li>
          <li>
            Presidential elections
          </li>
          <li>
            Abolition of slavery
          </li>
          <li>
            Civil rights
          </li>
          <li>
            Black suffrage
          </li>
          <li>
            Income taxes
          </li>
          <li>
            Senatorial elections
          </li>
          <li>
            Prohibition of liquor
          </li>
          <li>
            Women's suffrage
          </li>
          <li>
            Terms of office
          </li>
          <li>
            Repeal of Prohibition
          </li>
          <li>
            Term Limits for the Presidency
          </li>
          <li>
            Washington, D.C., suffrage
          </li>
          <li>
            Abolition of poll taxes
          </li>
          <li>
            Presidential succession
          </li>
          <li>
            18-year-old suffrage
          </li>
          <li>
            Congressional pay raises
          </li>
        </ol>
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="tab-panel--disabled"
        disabled=""
        hidden=""
        role="tabpanel"
      />
    </div>
  </div>
</div>
`;
